Add streaming I/O parser

radiohead opened the pull request:
### What

* Add streaming I/O parser implementation
* Use streaming I/O parser in CLI client
* Add mutex protection to all mutating listener methods
* Add new Makefile target for verifying the parse against "golden" output

### Why

* As per @Kaali request, we needed a new parser that implements `io.ReadWriteCloser` interface
* Mutex protection ensures that listener is safe for concurrent use
* CLI client switched to new parser for better speed (~15% faster than old)
* Verification step adds some safety net against parser being completely broken

Commit:
Add streaming I/O parser

* Add mutex protection to all mutating listener methods
* Use io.Copy in the CLI client

Commit:
Add a step to verify parser against "golden" output

Commit:
Add usage example to README

Commit:
Add separate "golden" build for verification

Commit:
Remove redundant comment from Makefile

radiohead commented:
@Kaali FWIW, I've spent an hour switching to `bytes.Buffer`, instead of doing the buffering manually and it made the code simpler, but a) there seem to be corner cases with my implementation and b) it's slightly slower than this one.

You can check the `bytes.Buffer` version here - https://github.com/speechly/nlu-example-parser/commit/4ba4c3310b08616a41071697d38b4a0072ebaea6.

I will investigate that branch later, when I have some spare time.

Kaali commented on pkg/parser/io_parser.go:
```diff
@@ -0,0 +1,152 @@
+
+	// If somehow we have exactly cap(b) in output buffer, just copy and return.
+	if outbuflen == bufcap {
+		copy(b, p.outputBuffer)
```
Isn't it wrong to use `cap(b)` when calculating these copies? The [io.Reader](https://golang.org/pkg/io/#Reader) API specifies that `Read reads up to len(p) bytes into p.` Where `p` is our `b`. Same also applies to the builtin `copy`.

So if I understand correctly, if the `bufcap` matched `outbuflen` but `len(b) == 0` then nothing is copied, but the code thinks something has been copied.

If I understand the copy section before the for-loop below. This is going to fill `b` with buffered data as much as it can. And when the buffer runs out it reads more. So could the implementation be more akin to:

```
written := copy(b, p.outputBuffer)
p.outputBuffer = p.outputBuffer[n:] // maybe copying the buffer to the start of the underlying slice buffer as at L52
if written == len(b) {
  return n, nil
}
```

Same buf vs len discussion applies to all code below.

radiohead replied:
> Isn't it wrong to use cap(b) when calculating these copies? The io.Reader API specifies that Read reads up to len(p) bytes into p. Where p is our b. Same also applies to the builtin copy.

Hey, that's a good point, and a simple fix. In reality, 99% of the cases are such that `cap(b) = len(b)`, but it is not a guarantee, and as you said, the interface explicitly requires relying on length, rather than capacity 👍 

> If I understand the copy section before the for-loop below ...

Yes, this should work!

radiohead replied:
Updated.

Kaali commented on pkg/parser/io_parser.go:
```diff
@@ -0,0 +1,152 @@
+		// We can fit more than we have available for writing.
+		// Write as much as we can and continue to the next iteration.
+		copy(b[written:], data)
+		written = written + datalen
```
Shouldn't this use the return value of copy to really know how much was copied?

radiohead replied:
Not necessarily, because we already know ahead of time how much will be copied. But this can be implemented in the same fashion as https://github.com/speechly/nlu-example-parser/pull/14#discussion_r388093987 proposes.

radiohead replied:
Updated.

Kaali commented on pkg/parser/io_parser.go:
```diff
@@ -0,0 +1,152 @@
+		// then just append it to our "current line" buffer and return.
+		if linesep == -1 {
+			p.inputBuffer = append(p.inputBuffer, b...)
+			return written + buflen, nil
```
For clarity this should be just `buflen`

radiohead replied:
No, this will not work, because `buflen` is not the full length of the buffer, but just the slice of `b`, that current iteration of the loop is working with.

Kaali replied:
Ah yeah, I kinda missed that this is in a loop. Which also confused me in the comment below. Both are resolved for me.

Kaali commented on pkg/parser/io_parser.go:
```diff
@@ -0,0 +1,152 @@
+		// Head contains the bytes [0 : newlineIndex]
+		// Re-slice the buffer to make sure we are progressively moving forward on it.
+		head := b[:linesep]
+		b = b[linesep+newlineLen:]
```
What does this do?  `b` is not used again. And without returning `b` the slice doesn't really change in the caller. And if it did, it would break [io.Writer](https://golang.org/pkg/io/#Writer) specification that slice data must not change (well data data wouldn't change, just the slice indices).

radiohead replied:
> What does this do?

As the comment states: "Re-slice the buffer to make sure we are progressively moving forward on it.". In other words, this will change `b` into a subset of `b`, starting at `linesep+newlineLen`, e.g. if `b = []byte{ 0, 1, 2, 3, 4}` and offset is `2`, after re-slicing, `b = []byte{2, 3, 4}`.

> b is not used again.

It is used, in the next iteration of the for-loop. This is done so that if `b` contains multiple "lines" (byte sequences, delimited by newline), in this loop we progressively move forward on `b` by searching and reading the next line in each iteration of the loop.

Kaali commented on pkg/parser/io_parser.go:
```diff
@@ -0,0 +1,152 @@
+			// Update input buffer, get the line to be parsed and zero out the input buffer.
+			p.inputBuffer = append(p.inputBuffer, head...)
+			line = string(p.inputBuffer)
+			p.inputBuffer = p.inputBuffer[:0]
```
Shouldn't the input buffer read the rest of `b`? By  [io.Writer](https://golang.org/pkg/io/#Writer) all data should be consumed or error raised.

radiohead replied:
> Shouldn't the input buffer read the rest of b?

No, because the rest of b is read directly to `parser.Parse`. This branch of code is for the case when `Write` is invoked N (N > 1) times, but we only get to the newline char at the last invocation.

Consider the following example:

```
1. parser.Write("str")
2. parser.Write("ing\n")
```

In [1], the bytes are stored in `inputBuffer`, in [2], we detect a newline and then we have to check the `inputBuffer`. In this example, it does contain something, so we concat it with the `head` (i.e. the chars until the newline) and then send the whole line to the parser.

If `b` would've been e.g. `"ing\nstr"` in [2], then the after re-slicing on L132, `b` would only contain `str`, which will be handled on the next loop iteration on L124.

Kaali replied:
Yeah, the same as above, missed the loop 👍

Commit:
Use len(b) instead of cap(b) in Read

Commit:
Return an error if input buffer is zero-length

Commit:
Use json.Encoder and bytes.Buffer for JSON serialization

Commit:
Simplify output buffer handling

Commit:
Use buffered reading in CLI instead of io.Copy

radiohead commented:
@Kaali I've addressed your comments, but added some extra changes, after looking at https://github.com/speechly/s2t-grinder/pull/119 by @nori266 - it seems like we actually do need each JSON to be newline-separated.

The simplest way to achieve that without losing performance was to use `json.Encoder.Encode`, but since it requires an `io.Writer`, I've opted into using an extra `bytes.Buffer` for that.

Now, I realise it introduces redundancy to the `outputBuffer`, but I would rather keep it that way for now and address that in a separate PR.

@nori266 - FYI I've 1) updated the `golden.json` to include a newline and 2) changed CLI parser to use buffered read / writes, so that those can be controlled by `-buffer_size_mb` flag.

radiohead commented:
Oh, and trying to pass 0-length buffers to `parser.Read` / `parser.Write` will now return an error - I've accidentally done that myself and it just silently failed, so IMO it's a necessary addition.

Commit:
Undo redundant changes to listener
